pkg/engine/internal/workflow/workflow_planner.go

```diff
@@ -0,0 +1,408 @@
+// have exactly one root node, or if the physical plan cannot be partitioned.
+func planWorkflow(plan *physical.Plan) (dag.Graph[*Task], error) {
+	root, err := plan.Root()
+	if err != nil {
+		return dag.Graph[*Task]{}, err
+	}
+
+	planner := &planner{
+		physical: plan,
+
+		streamWriters: make(map[*Stream]*Task),
+		streamReaders: make(map[*Stream]*Task),
+	}
+	if err := planner.Process(root); err != nil {
+		return dag.Graph[*Task]{}, err
+	}
+
+	return planner.graph, nil
+}
+
+// Process builds a set of tasks from a root physical plan node. Built tasks are
+// added to p.graph.
+func (p *planner) Process(root physical.Node) error {
+	_, err := p.processNode(root, true)
+	return err
+}
+
+// processNode builds a set of tasks from the given node. splitOnBreaker
+// indicates whether pipeline breaker nodes should be split off into their own
+// task.
+//
+// The reuslting task is the task immediately produced by node, which callers
+// can use to add edges. All tasks, including those produced in recursive calls
+// to processNode, are added into p.Graph.
+func (p *planner) processNode(node physical.Node, splitOnBreaker bool) (*Task, error) {
+	var (
+		// taskPlan is the in-progress physical plan for an individual task.
+		taskPlan dag.Graph[physical.Node]
+
+		sources = make(map[physical.Node][]*Stream)
+
+		// childrenTasks is the slice of immediate Tasks produced by processing
+		// the children of node.
+		childrenTasks []*Task
+	)
+
+	// Immediately add the node to the task physical plan.
+	taskPlan.Add(node)
+
+	var (
+		stack     = make(stack[physical.Node], 0, p.physical.Len())
+		visited   = make(map[physical.Node]struct{}, p.physical.Len())
+		nodeTasks = make(map[physical.Node][]*Task, p.physical.Len())
+	)
+
+	stack.Push(node)
+	for stack.Len() > 0 {
+		next := stack.Pop()
+		if _, ok := visited[next]; ok {
+			// Ignore nodes that have already been visited, which can happen
+			// if a node has multiple parents.
+			continue
+		}
+		visited[next] = struct{}{}
+
+		for _, child := range p.physical.Children(next) {
+			// NOTE(rfratto): We may have already seen child before (if it has
+			// more than one parent), but we want to continue processing to
+			// ensure we update sources properly and retain all relationships
+			// within the same graph.
+
+			switch {
+			case splitOnBreaker && isPipelineBreaker(child):
+				childTasks, found := nodeTasks[child]
+				if !found {
+					// Split the pipeline breaker into its own set of tasks.
+					task, err := p.processNode(child, splitOnBreaker)
+					if err != nil {
+						return nil, err
+					}
+					childrenTasks = append(childrenTasks, task)
+					nodeTasks[child] = append(nodeTasks[child], task)
+					childTasks = nodeTasks[child]
+				}
+
+				// Create one unique stream for each child task so we can
+				// receive output from them.
+				for _, task := range childTasks {
+					stream := &Stream{ULID: ulid.Make()}
```

Comment:
It _seems_ it's possible to use `Stream` as a value which should reduce the number of allocations. Does it make sense to do so or do you have other plans for `Stream`?